ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/LedgerDB.hs

```diff
@@ -1,82 +1,1987 @@
+{-# LANGUAGE BangPatterns               #-}
+{-# LANGUAGE CPP                        #-}
+{-# LANGUAGE ConstraintKinds            #-}
+{-# LANGUAGE DataKinds                  #-}
+{-# LANGUAGE DeriveAnyClass             #-}
+{-# LANGUAGE DeriveGeneric              #-}
+{-# LANGUAGE DerivingVia                #-}
+{-# LANGUAGE FlexibleContexts           #-}
+{-# LANGUAGE GADTs                      #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase                 #-}
+{-# LANGUAGE NamedFieldPuns             #-}
+{-# LANGUAGE NumericUnderscores         #-}
+{-# LANGUAGE RankNTypes                 #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+{-# LANGUAGE StandaloneDeriving         #-}
+{-# LANGUAGE TupleSections              #-}
+{-# LANGUAGE TypeApplications           #-}
+{-# LANGUAGE TypeFamilies               #-}
+{-# LANGUAGE TypeOperators              #-}
+
+-- | The Ledger DB is responsible for the following tasks:
+--
+-- - __Maintaining the in-memory ledger state at the tip__: Maintaining the
+--     ledger state corresponding to the current tip in memory. When we try to
+--     extend our chain with a new block fitting onto our tip, the block must
+--     first be validated using the right ledger state, i.e., the ledger state
+--     corresponding to the tip. The current ledger state is needed for various
+--     other purposes.
+--
+-- - __Maintaining the past \(k\) in-memory ledger states__: we might roll back
+--     up to \(k\) blocks when switching to a more preferable fork. Consider the
+--     example below:
+--
+--     <<docs/haddocks/ledgerdb-switch.svg>>
+--
+--     Our current chain's tip is \(C_2\), but the fork containing blocks
+--     \(F_1\), \(F_2\), and \(F_3\) is more preferable. We roll back our chain
+--     to the intersection point of the two chains, \(I\), which must be not
+--     more than \(k\) blocks back from our current tip. Next, we must validate
+--     block \(F_1\) using the ledger state at block \(I\), after which we can
+--     validate \(F_2\) using the resulting ledger state, and so on.
+--
+--     This means that we need access to all ledger states of the past \(k\)
+--     blocks, i.e., the ledger states corresponding to the volatile part of the
+--     current chain. Note that applying a block to a ledger state is not an
+--     invertible operation, so it is not possible to simply /unapply/ \(C_1\)
```

Comment:
It is not an invertible operation under the current design?

Reply:
That's right, but in the overall design of all the layers of the node.